{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Data import CodonTable\n",
    "import pandas as pd\n",
    "\n",
    "## 1 ##\n",
    "    \n",
    "    def get_sequences_from_file(fasta_fn):\n",
    "\n",
    "    sequence_data_dict = {}\n",
    "    \n",
    "    for record in SeqIO.parse(fasta_fn, \"fasta\"):\n",
    "    \n",
    "        description = record.description.split()\n",
    "        \n",
    "        species_name = description[1] + \" \" + description[2] \n",
    "        \n",
    "        sequence_data_dict[species_name] = record.seq \n",
    "             \n",
    "    return(sequence_data_dict)\n",
    "\n",
    "    \n",
    "    \n",
    "    ## 2 ##\n",
    "def translate_function(string_nucleotides): \n",
    "    mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"] \n",
    "    \n",
    "    aa_seq_string =\"\"\n",
    "    for i in range(0, len(string_nucleotides), 3):  \n",
    "    \n",
    "        codon = string_nucleotides[i:i+3]\n",
    "        if codon in mito_table.stop_codons:\n",
    "            break        \n",
    "            \n",
    "        aminoacid = mito_table.forward_table[codon] \n",
    "          \n",
    "        aa_seq_string += aminoacid        \n",
    "        \n",
    "    return(aa_seq_string)\n",
    "\n",
    "\n",
    "## 3 ##\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "def Alt_translate(string_nucleotides): \n",
    "\n",
    "    coding_dna = Seq(string_nucleotides) \n",
    "    \n",
    "    result = string_nucleotides.translate(table=\"Vertebrate Mitochondrial\", to_stop=True) \n",
    "    \n",
    "    AA_Sequence = str(result) \n",
    "    \n",
    "    return(AA_Sequence) \n",
    "  \n",
    "\n",
    "##### 4 #####\n",
    "def get_proportion_aa(aa_seq):\n",
    "   \n",
    "    charged = ['R','K','D','E'] \n",
    "    polar = ['Q','N','H','S','T','Y','C','M','W']\n",
    "    hydrophobic = ['A','I','L','F','V','P','G']\n",
    "    c_count=0\n",
    "    p_count=0\n",
    "    h_count=0\n",
    "    \n",
    "    for aa in str(aa_seq):\n",
    "        if aa in charged:\n",
    "            c_count += 1\n",
    "            \n",
    "        elif aa in polar:\n",
    "            p_count += 1\n",
    "            \n",
    "        elif aa in hydrophobic:\n",
    "            h_count += 1\n",
    "            \n",
    "    proportion_type = [c_count/len(aa_seq), p_count/len(aa_seq), h_count/len(aa_seq)]\n",
    "    \n",
    "    return proportion_type\n",
    "\n",
    "\n",
    "#Main function\n",
    "\n",
    "cytb_seqs = get_sequences_from_file(\"bears_cytb.fasta\") \n",
    "\n",
    "bear_df = pd.read_csv(\"bears_data.csv\") \n",
    "\n",
    "species_list = list(bear_df.species)\n",
    "\n",
    "for key, value in cytb_seqs.items():\n",
    "    aa_seq =Alt_translate(str(value)) \n",
    "    \n",
    "    proportion = get_proportion_aa(aa_seq) \n",
    "    \n",
    "    bear_df.loc[bear_df.species==key, 'charged'] = proportion[0]\n",
    "    \n",
    "    bear_df.loc[bear_df.species==key, 'polar'] = proportion[1]\n",
    "    \n",
    "    bear_df.loc[bear_df.species==key, 'hydrophobic'] = proportion[2] \n",
    "    \n",
    "    #5#\n",
    "\n",
    "import seaborn as sea\n",
    "\n",
    "bar_chart = sea.barplot(x = 'species', y = 'mass', data=bear_df) \n",
    "\n",
    "bar_chart.set_xticklabels(labels = bear_df.species, rotation=30, horizontalalignment='right') \n",
    "\n",
    "\n",
    "\n",
    "## 6 ##\n",
    "\n",
    "#The largest bear species is the Ursus spelaeus.\n",
    "\n",
    "#The other thing that is interesting about this species is that the spelaeus does not demonstrate a significantly difference gene sequence than the other species. Huh. \n",
    "\n",
    "## 7 ##\n",
    "\n",
    "import numpy as num\n",
    "import matplotlib.pyplot as mpl\n",
    "ind = num.arange(len(bear_df))\n",
    "p1= mpl.bar(ind, bear_df.charged)\n",
    "p2= mpl.bar(ind, bear_df.polar, bottom = bear_df.charged)\n",
    "p3= mpl.bar(ind, bear_df.hydrophobic, bottom = bear_df.charged + bear_df.polar)\n",
    "mpl.xlabel('Ursus Species')\n",
    "mpl.xticks(ind, bear_df.species, rotation=30, horizontalalignment='right')\n",
    "mpl.ylabel('Proportion')\n",
    "mpl.yticks(np.arange(0, 1.1, 0.1))\n",
    "mpl.legend((p1[0], p2[0], p3[0]), ('Charged', 'Polar', 'Hydrophobic'), bbox_to_anchor=(1.01, 1.01))\n",
    "mpl.show()\n",
    "\n",
    "\n",
    "## 8 ##\n",
    "\n",
    "bear_df.to_csv('bears_mass_cytb.csv')\n",
    "\n",
    "\n",
    "#bonus\n",
    "\n",
    "Bonus = get_sequences_from_file(\"bears_cytb.fasta\") \n",
    "\n",
    "nuc_count=pd.DataFrame\n",
    "\n",
    "for data,value in Bonus.items():\n",
    " \n",
    "    G=0\n",
    "    A=0\n",
    "    C=0\n",
    "    T=0\n",
    "    \n",
    "    for j in range(0,len(value)):\n",
    "    \n",
    "        nuc=value[j]\n",
    "        if nuc=='G':\n",
    "            G=G+1\n",
    "       \n",
    "        elif nuc=='A':\n",
    "            A=A+1\n",
    "     \n",
    "        elif nuc=='T':\n",
    "            T=T+1\n",
    "\n",
    "        else:\n",
    "            C=C+1 \n",
    "    print(data)\n",
    "    print(\"G={}, A={}, C={}, T={}\".format(G,A,C,T))\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
